src/interfaces/records/handlers/records-upload.ts

```diff
@@ -0,0 +1,196 @@
+
+type RecordsUploadMessageVariant = RecordsUploadCompleteMessage | RecordsUploadPartMessage | RecordsUploadStartMessage;
+
+export class RecordsUploadHandler implements MethodHandler {
+
+  constructor(private didResolver: DidResolver, private messageStore: MessageStore, private uploadStore: UploadStore) { }
+
+  public async handle({
+    tenant,
+    message,
+    dataStream
+  }): Promise<MessageReply> {
+    const incomingMessage = message as RecordsUploadMessageVariant;
+
+    let recordsUpload: RecordsUpload;
+    try {
+      recordsUpload = await RecordsUpload.parse(incomingMessage);
+    } catch (e) {
+      return new MessageReply({
+        status: { code: 400, detail: e.message }
+      });
+    }
+
+    // authentication & authorization
+    try {
+      await authenticate(incomingMessage.authorization, this.didResolver);
+      await recordsUpload.authorize(tenant, this.messageStore);
+    } catch (e) {
+      return new MessageReply({
+        status: { code: 401, detail: e.message }
+      });
+    }
+
+    // get existing upload start messages matching the `recordId`
+    const query = {
+      interface : DwnInterfaceName.Records,
+      method    : DwnMethodName.Upload,
+      recordId  : incomingMessage.recordId
+    };
+    const existingMessages = await this.messageStore.query(tenant, query) as RecordsUploadMessageVariant[];
+
+    try {
+      const existingUploadComplete = await RecordsUpload.getUploadComplete(existingMessages);
+      if (existingUploadComplete) {
+        return new MessageReply({
+          status: { code: 400, detail: 'upload already complete' }
+        });
+      }
+    } catch (e) {
+      // ignore error if upload complete cannot be found
+    }
+
+    const newMessageIsUploadStart = await recordsUpload.isUploadStart();
+    const newMessageIsUploadComplete = await recordsUpload.isUploadComplete();
+
+    const parts = new Set<number>();
+
+    try {
+      const existingUploadStart = await RecordsUpload.getUploadStart(existingMessages);
+      if (existingUploadStart) {
+        if (newMessageIsUploadStart) {
+          return new MessageReply({
+            status: { code: 400, detail: 'cannot start an upload more than once' }
+          });
+        }
+
+        RecordsUpload.verifyEqualityOfImmutableProperties(existingUploadStart, incomingMessage);
+      }
+
+      for (const existingMessage of existingMessages) {
+        RecordsUpload.verifyExclusivityOfUniqueProperties(existingMessage, incomingMessage);
+
+        if (newMessageIsUploadComplete) {
+          if (existingMessage.descriptor.state === DwnStateName.Part) {
+            const uploadPart = existingMessage as RecordsUploadPartMessage;
+
+            parts.add(uploadPart.descriptor.index);
+          }
+        }
+      }
+    } catch (e) {
+      return new MessageReply({
+        status: { code: 400, detail: e.message }
+      });
+    }
+
+    if (newMessageIsUploadComplete) {
```

Comment:
We check that we have unique indexed parts up to `uploadComplete.descriptor.count`. We do not do this index check when we are handling a `RecordsUploadPart`. I'm guessing the reason is that we may receive parts out of order? If so, is it also possible that we may receive the `Complete` message before we finish receiving the `Part`s?